python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,225 @@
+
+    cdef struct forest_params_t:
+        dense_node_t* nodes
+        int depth
+        int ntrees
+        int cols
+        algo_t algo
+        output_t output
+        float threshold
+        pass
+
+    cdef struct treelite_params_t:
+        algo_t algo
+        bool output_class
+        float threshold
+
+    cdef void dense_node_init(dense_node_t*,
+                              float,
+                              float,
+                              int,
+                              bool,
+                              bool)
+
+    cdef void dense_node_decode(dense_node_t*,
+                                float*,
+                                float*,
+                                int*,
+                                bool*,
+                                bool*)
+
+    cdef void init_dense(cumlHandle& handle,
+                         forest_t*,
+                         forest_params_t*)
+
+    cdef void free(cumlHandle& handle,
+                   forest_t)
+
+    cdef void predict(cumlHandle& handle,
+                      forest_t,
+                      float*,
+                      float*,
+                      size_t)
+
+    cdef forest_t from_treelite(cumlHandle& handle,
+                                forest_t*,
+                                ModelHandle,
+                                treelite_params_t*)
+
+cdef class FIL_impl():
+
+    cpdef object handle
+    cdef forest_t* forest_pointer
+    cdef forest_t forest_data
+    cdef ModelHandle* model_ptr
+    cdef object depth
+    cdef object num_trees
+    cdef object algo
+    cdef object output
+    cdef object threshold
+    cdef object n_cols
+    cdef object dtype
+    cdef object verbose
+
+    def __cinit__(self, depth=8, n_estimators=50,
+                  output=0, algo=0,
+                  threshold=0.0,
+                  handle=None, verbose=False):
+
+        self.depth = depth
+        self.num_trees = n_estimators
+        self.output = output
+        self.algo = algo
+        self.threshold = threshold
+        self.handle = handle
+        self.verbose = verbose
+        self.forest_pointer = NULL
+
+    def predict(self, X):
+        cdef uintptr_t X_ptr
+        X_m, X_ptr, n_rows, _, X_dtype = \
+            input_to_dev_array(X, order='C')
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        preds = np.ones(n_rows, dtype=np.float32)
+        cdef uintptr_t preds_ptr
+        preds_m, preds_ptr, _, _, _ = \
+            input_to_dev_array(preds)
+        print( " dtype of X : ", X_dtype)
+        #cdef forest_t forest_data = <forest_t> &self.forest_pointer
+        predict(handle_[0],
+                self.forest_data,
+                <float*> preds_ptr,
+                <float*> X_ptr,
+                <size_t> n_rows)
+        self.handle.sync()
+        # synchronous w/o a stream
+        preds = preds_m.copy_to_host()
+        return preds
+
+    def free(self):
```

Comment:
This should be called in the standard Python function that works as an object destructor, e.g. `__del__()` or `__exit__()`.